### ISSUE TITLE
Controlling the global step in TrainResult.log and EvalResult.log

### ISSUE THREAD

**Issue by shalgi-beyond** [PyTorchLightning/pytorch-lightning#3339]
I can find a way to control the global step of logging metrics to TrainResult using the log (or log_dict) functions?
What is the proper way to use these functions so the logged metrics will show in tensorboard once per epoch?
Currently, the steps showing on my tensorboard are (63*i - 1) for i=1 and so on.
This is my training_step function (validation_step is similar using pl.EvalResult)
    
    def training_step(self, batch, batch_idx) -> pl.TrainResult:
        x, mask = batch
        pred = self(x)
        loss = self.loss_function(pred, mask)
        result = pl.TrainResult(loss)
        result.log("Trainer/cross_entropy_loss", self.cross_entropy_loss(pred, mask))
        return result

I tried to set most of `TrainResult.log` parameters manually (like on_epoch, logger, sync_dist, reduce_fx and so on)

 - OS: Ubuntu 18.04.4
Nvidia driver version: 440.33.01
CUDA versions available: cuda-10.0 cuda-10.1 cuda-10.2 cuda-9.0 cuda-9.2
Default CUDA version is 10.0
 - torch==1.5.1
 - pytorch-lightning==0.9.0

Thank you a lot in advance

**Comment by ydcjeff** [comment]
you can try 
- `result.log(on_step=True)` to only log once per batch
- `result.log(on_epoch=True)` to only log once per epoch
- `result.log(on_step=True, on_epoch=True)` to log once per batch and once per epoch

**Comment by shalgi-beyond** [comment]
@ydcjeff thank you for the quick answer, I actually did try all the above and the step showing in tensorboard is not 0,1,2 and so on but it starts with a bigger number and bigger jumps

**Comment by shalgi-beyond** [comment]
@ydcjeff yes of course!
`import os
import torch
import torch.nn.functional as F
from torchvision.datasets import MNIST
from torchvision import transforms
from torch.utils.data import DataLoader
import pytorch_lightning as pl
from torch.utils.data import random_split
class LitModel(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.layer_1 = torch.nn.Linear(28 * 28, 128)
        self.layer_2 = torch.nn.Linear(128, 10)
    def forward(self, x):
        x = x.view(x.size(0), -1)
        x = self.layer_1(x)
        x = F.relu(x)
        x = self.layer_2(x)
        return x
    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=1e-3)
        return optimizer
    def training_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        loss = F.cross_entropy(y_hat, y)
        result = pl.TrainResult(loss)
        result.log('train_loss', loss, on_step=True, on_epoch=False)
        return result

    def validation_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        loss = F.cross_entropy(y_hat, y)
        result = pl.EvalResult(checkpoint_on=loss)
        result.log('val_loss', loss, on_step=True, on_epoch=False)
        return result
    def test_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        loss = F.cross_entropy(y_hat, y)
        result = pl.EvalResult()
        result.log('test_loss', loss, on_step=True, on_epoch=False)
        return result
# dataloaders
dataset = MNIST(os.getcwd(), download=True, transform=transforms.ToTensor())
train, val = random_split(dataset, [55000, 5000])
train_loader = DataLoader(train)
val_loader = DataLoader(val)
# init model
model = LitModel()
# most basic trainer, uses good defaults (auto-tensorboard, checkpoints, logs, and more)
trainer = pl.Trainer(gpus=1, log_save_interval=1)
trainer.fit(model, train_loader, val_loader)`

**Comment by shalgi-beyond** [comment]
@ydcjeff  Yes of course!
    import os
    import torch
    import torch.nn.functional as F
    from torchvision.datasets import MNIST
    from torchvision import transforms
    from torch.utils.data import DataLoader
    import pytorch_lightning as pl
    from torch.utils.data import random_split
    class LitModel(pl.LightningModule):
        def __init__(self):
            super().__init__()
            self.layer_1 = torch.nn.Linear(28 * 28, 128)
            self.layer_2 = torch.nn.Linear(128, 10)
        def forward(self, x):
            x = x.view(x.size(0), -1)
            x = self.layer_1(x)
            x = F.relu(x)
            x = self.layer_2(x)
            return x
        def configure_optimizers(self):
            optimizer = torch.optim.Adam(self.parameters(), lr=1e-3)
            return optimizer
        def training_step(self, batch, batch_idx):
            x, y = batch
            y_hat = self(x)
            loss = F.cross_entropy(y_hat, y)
            result = pl.TrainResult(loss)
            result.log('train_loss', loss, on_step=True, on_epoch=False)
            return result
        
        def validation_step(self, batch, batch_idx):
            x, y = batch
            y_hat = self(x)
            loss = F.cross_entropy(y_hat, y)
            result = pl.EvalResult(checkpoint_on=loss)
            result.log('val_loss', loss, on_step=True, on_epoch=False)
            return result
        def test_step(self, batch, batch_idx):
            x, y = batch
            y_hat = self(x)
            loss = F.cross_entropy(y_hat, y)
            result = pl.EvalResult()
            result.log('test_loss', loss, on_step=True, on_epoch=False)
            return result
    # dataloaders
    dataset = MNIST(os.getcwd(), download=True, transform=transforms.ToTensor())
    train, val = random_split(dataset, [55000, 5000])
    train_loader = DataLoader(train)
    val_loader = DataLoader(val)
    # init model
    model = LitModel()
    # most basic trainer, uses good defaults (auto-tensorboard, checkpoints, logs, and more)
    trainer = pl.Trainer(gpus=1, log_save_interval=1)
    trainer.fit(model, train_loader, val_loader)

**Comment by shalgi-beyond** [comment]
also, I get a weird "epoch" metric on my tensorboard.
![image](https://user-images.githubusercontent.com/64365020/92130405-5cf88d80-ee0d-11ea-97f5-16a23f15eded.png)
I do not have any logic that creates this or even the word "epoch" in my code. (other than on_epoch=True on some experiments)

**Comment by ydcjeff** [comment]
Hi @shalgi-beyond setting `trainer = pl.Trainer(gpus=1, log_save_interval=1, row_log_interval=1)` would do the trick.

**Comment by awaelchli** [comment]
@shalgi-beyond the plot with the "epoch" is added by lightning so that you know how many global steps correspond to one epoch. For example, in your case 700 steps = 20 epochs.
The global_step is not meant to be controlled by the user. Every row of data you log corresponds to one step. This is TensorBoard behaviour, not Lightning. 

We typically don't want to log every batch, that can be problematic, so we set the row_log_interval to a higher number. In rare cases when you have fewer batches than this number, you will see jumps.

**Comment by shalgi-beyond** [comment]
@ydcjeff thank you for the tip!
@awaelchli I understand that sometimes the user is not supposed to have control over internal details, but I still don't understand whether there is a way to log my metrics once per epoch, and display the TB graphs accordingly (so that the x axis is epochs).
This usage seems standard to me but I can't manage to create this using lightning currently.
I thought lightning would accumulate the results per epoch and write to TB just the average of the whole epoch, once when it ends.
Thank you a lot!

**Comment by s-rog** [comment]
#3228 what you're asking for has been looked into (epoch on x axis) but it would be problematic with metrics and other loggers and thus has not been implemented atm.
